### PR TITLE
elmPackages.elm-pages: init at 2.1.11

### PR DESCRIPTION
**Pull request by jali-clarke**
###### Description of changes

Inits `elmPackages.elm-pages` at version `2.1.11`.  Involves a regeneration of `node-packages.nix` as a result of running `generate-node-packages.sh`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by turboMaCk — APPROVED** [review]
nixpkgs-review build ok

<details>

12 packages built:
elmPackages.create-elm-app elmPackages.elm-analyse elmPackages.elm-coverage elmPackages.elm-doc-preview elmPackages.elm-live elmPackages.elm-optimize-level-2 elmPackages.elm-pages elmPackages.elm-review elmPackages.elm-test elmPackages.elm-upgrade elmPackages.elm-verify-examples elmPackages.elm-xref

</details>

**Comment by turboMaCk** [comment]
@jali-clarke is there anything left to do (doesn't seem like it) or are you just waiting for build to pass before marking this PR as ready?

**Comment by jali-clarke** [comment]
@turboMaCk we noticed that some of the files created by `elm-pages` are actually copied from the nix store and as a result are read-only.  I'll push a patch shortly to address this and then this will be ready to go.

**Comment by turboMaCk** [comment]
I wonder can we upstream this patch to elm-pages?

**Comment by jali-clarke** [comment]
Will attempt to do so; lemme get back to you on that

**Comment by jali-clarke** [comment]
@turboMaCk So it seems that the current `master` of `elm-pages` is broken (i.e. the executable doesn't work, regardless of whether it's packaged with `nix`) and there does not seem to be a backport process.  Would it be ok in this situation to accept the PR as-is?  Or is there a preferable course of action that I should take?

File: pkgs/development/compilers/elm/default.nix [hunk 1]
```diff
@@ -212,6 +212,31 @@ in lib.makeScope pkgs.newScope (self: with self; {
...
+            wrapProgram $out/bin/elm-pages --prefix PATH : ${
+              with pkgs.elmPackages; lib.makeBinPath [ elm elm-review ]
+            }
+          '';
+
+          meta = with lib; nodePkgs."elm-pages".meta // {
+            description = "A statically typed site generator for Elm.";
+            homepage = "https://github.com/dillonkearns/elm-pages";
+            license = licenses.bsd3;
+            maintainers = [ maintainers.turbomack ];
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/default.nix:235]: don't you want to also add yourself as a maintainer?

File: pkgs/development/compilers/elm/default.nix [hunk 2]
```diff
@@ -212,6 +212,31 @@ in lib.makeScope pkgs.newScope (self: with self; {
...
 
+      elm-pages = nodePkgs."elm-pages".overrideAttrs (
+        old: {
+          buildInputs = old.buildInputs ++ [ pkgs.makeWrapper ];
+
+          # can't use `patches = [ <patch_file> ]` with a nodePkgs derivation;
+          # need to patch in one of the build phases instead
+          preFixup = ''
+            patch $out/lib/node_modules/elm-pages/generator/src/codegen.js ${./packages/elm-pages-fix-read-only.patch}
+          '';
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/default.nix:223]: I think you can use  `patches`. It takes list of patch files to apply.

**Review by turboMaCk — CHANGES_REQUESTED** [review]
Hmm I think there are few things we should do:

- [x] open issue in elm-pages with the issue
- [x] add link to the issue above `patches` in derivation so we can track progress upstream
- [ ] commits should me squashed into single commit before merge

**Comment by jali-clarke** [comment]
Issue created @ https://github.com/dillonkearns/elm-pages/issues/305 to track the read-only problem.

File: pkgs/development/compilers/elm/default.nix [hunk 3]
```diff
@@ -212,6 +212,32 @@ in lib.makeScope pkgs.newScope (self: with self; {
...
+      elm-pages = nodePkgs."elm-pages".overrideAttrs (
+        old: {
+          buildInputs = old.buildInputs ++ [ pkgs.makeWrapper ];
+
+          # can't use `patches = [ <patch_file> ]` with a nodePkgs derivation;
+          # need to patch in one of the build phases instead.
+          # see upstream issue https://github.com/dillonkearns/elm-pages/issues/305 for dealing with the read-only problem
+          preFixup = ''
+            patch $out/lib/node_modules/elm-pages/generator/src/codegen.js ${./packages/elm-pages-fix-read-only.patch}
+          '';
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/default.nix:224]:
```suggestion
          patches = [ ''./packages/elm-pages-fix-read-only.patch'' ];
```

File: maintainers/maintainer-list.nix [hunk 4]
```diff
@@ -14671,4 +14671,10 @@
     github = "snpschaaf";
     githubId = 105843013;
   };
+  jali-clarke = {
+    email = "jinnah.ali-clarke@outlook.com";
+    name = "Jinnah Ali-Clarke";
+    github = "jali-clarke";
+    githubId = 17733984;
+  };
```
Comment (turboMaCk) [inline maintainers/maintainer-list.nix:14679]: Sorry for making this a bit more complicated but according to guidelines adding of new maintainer should be done in separate commit. This means that this PR should contain 2 commits.

- add jali-clarke to maintainers list
- elmPackages.elm-pages: init at 2.1.11

File: pkgs/development/compilers/elm/default.nix (hunk 3 above)
Comment (jali-clarke) [reply pkgs/development/compilers/elm/default.nix:224]: I did try this when I started this work, but this error appears in the build logs:
```
patching sources
applying patch /nix/store/zw27jx69p2m0a5a3cy89nk2rsknh9c5r-elm-pages-fix-read-only.patch
can't find file to patch at input line 5
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/generator/src/codegen.js b/generator/src/codegen.js
|index a0ef8806e..fe00ac904 100644
|--- a/generator/src/codegen.js
|+++ b/generator/src/codegen.js
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
```

Lemme poke at it a bit, will push a commit when I have `patches` working (or I give up)

Comment (turboMaCk) [reply pkgs/development/compilers/elm/default.nix:224]: it might be the case that node2nix builder doesn't support this. I haven't tried it. In that case it would be fine to keep it as is.

Comment (turboMaCk) [reply pkgs/development/compilers/elm/default.nix:224]: I believe there is even `patchPhase` but again I'm now not sure if it works with derivations from node2nix. I don't think postInstall is the right place for applying patches even though in this case it would be possible. I would try `patchPhase` and if that doesn't work I would keep it in `preFixup`

Comment (jali-clarke) [reply pkgs/development/compilers/elm/default.nix:224]: `patchPhase` doesn't work either, seeing the following repeated ad infinitum:

```
       > patch: **** Can't create file /nix/store/nwk4ir8gxc2fc9ma3xg82vzgqf7dzkx5-elm-pages-2.1.11/lib/node_modules/elm-pages/generator/src/codegen.js.orig : Too many open files
       > patch: **** Can't create file /nix/store/nwk4ir8gxc2fc9ma3xg82vzgqf7dzkx5-elm-pages-2.1.11/lib/node_modules/elm-pages/generator/src/codegen.js.orig : Too many open files
       > patch: **** Can't create file /nix/store/nwk4ir8gxc2fc9ma3xg82vzgqf7dzkx5-elm-pages-2.1.11/lib/node_modules/elm-pages/generator/src/codegen.js.orig : Too many open files
```

Will leave it in `preFixup` as you've advised.

**Review by turboMaCk — APPROVED** [review]
LGTM

**Comment by turboMaCk** [comment]
@GrahamcOfBorg eval elmPackages.elm-pages
